src/file/FileControllerNative.ts

```diff
@@ -67,44 +100,65 @@ export class FileControllerNative implements FileController {
 
 	/**
 	 * Temporary files are deleted afterwards in apps.
-	 *
-	 * TODO this could probably just use this.doDownload. Temporary files are not being cleaned up on android
 	 */
 	async downloadAll(tutanotaFiles: Array<TutanotaFile>): Promise<void> {
-		const downloadAll = async <T>(downloadFile: (file: TutanotaFile) => Promise<T>, processDownloadedFiles: (downloadedFiles: T[]) => Promise<unknown>) => {
-			const downloadedFiles: Array<T> = []
-			for (const file of tutanotaFiles) {
-				try {
-					const downloadedFile = await downloadFile(file)
-					downloadedFiles.push(downloadedFile)
-				} catch (e) {
-					await handleDownloadErrors(e, (msg) => Dialog.message(() => lang.get(msg) + " " + file.name))
-				}
-			}
-			await processDownloadedFiles(downloadedFiles)
-		}
-
+		const progress = stream(0)
 		if (isAndroidApp()) {
-			await downloadAll(
-				(file) => this.downloadAndDecryptInNative(file),
-				(files) => promiseMap(files, (file) => this.fileApp.putFileIntoDownloadsFolder(file.location)),
+			await guiDownload(
+				this.doDownload(
+					tutanotaFiles,
+					(file) => this.downloadAndDecryptInNative(file),
+					(files) => promiseMap(files, (file) => this.fileApp.putFileIntoDownloadsFolder(file.location)),
+					(files) => this.cleanUp(files),
+					progress,
+				),
+				progress,
 			)
 		} else if (isIOSApp()) {
-			await downloadAll(
-				(file) => this.downloadAndDecryptInNative(file),
-				(files) =>
-					promiseMap(files, async (file) => {
-						try {
-							await this.fileApp.open(file)
-						} finally {
-							await this.fileApp.deleteFile(file.location).catch((e) => console.log("failed to delete file", file.location, e))
-						}
-					}),
+			await guiDownload(
+				this.doDownload(
+					tutanotaFiles,
+					(file) => this.downloadAndDecryptInNative(file),
+					(files) =>
+						promiseMap(files, async (file) => {
+							try {
+								await this.fileApp.open(file)
```

Comment:
are we actually opening each file after download?

Reply:
apparently there is no such thing like a direct download in iOS. When clicking on an attachment it will open and then offer the option to download. That's why the "Save all" option is not available on iOS. I was also confused why this code exists but decided not to mess with it. Should I rather remove it? As far as I can tell this case can never happen